main.py

```diff
@@ -150,16 +151,36 @@ def get_active_rules():
         arguments = line.split()
         rule = parse_rule(arguments)
         if rule["name"] and rule["name"] not in dev:
+            rule["ip"] = get_interface_ip(rule["name"])
             rules.append(rule)
             dev.add(rule["name"])
             rules.sort(key=lambda x: x["name"])
     return rules
 
 
+def get_interfaces():
+    output = subprocess.check_output(["ip", "-o", "-4", "addr", "show"]).decode(sys.stdout.encoding)
+    interfaces = {}
+    for line in output.split('\n'):
+        if line:
+            parts = line.split()
+            iface = parts[1]
+            ip = parts[3].split('/')[0]
+            interfaces[iface] = ip
+    return interfaces
+
+
+def get_interface_ip(interface):
+    output = subprocess.check_output(["ip", "-o", "-4", "addr", "show"]).decode(sys.stdout.encoding)
```

Comment:
This call is duplicated from the other function. Can we extract it to a separate function to avoid the duplication.

Or since you already pass the interface, why not get the IP from there?

Reply:
please see my latest commit

Reply:
Could you try to find another solution than popen? I remember that the docs recommended to use run instead.